src/runtime_src/core/common/utils.cpp

```diff
@@ -183,24 +185,57 @@ unit_convert(size_t size)
 uint16_t
 bdf2index(const std::string& bdfstr, bool _inUserDomain)
 {
-  auto n = std::count(bdfstr.begin(), bdfstr.end(), ':');
-
-  std::stringstream s(bdfstr);
-  uint16_t dom = 0, b = 0, d = 0, f = 0;
-  char dummy;
-
-  if (n == 2)
-    s >> std::hex >> dom >> dummy;
-  s >> std::hex >> b >> dummy >> d >> dummy >> f;
-
-  if ((n != 1 && n != 2) || s.fail())
-    throw std::runtime_error("Bad BDF string '" + bdfstr + "'");
+  std::vector<std::string> tokens; 
+  boost::split(tokens, bdfstr, boost::is_any_of(":")); 
+  int radix = 16;
+  uint16_t bus = 0, dev = 0, func = std::numeric_limits<uint16_t>::max();
 
+  //throw an eror if no devices are present
   uint64_t devices = _inUserDomain ? xrt_core::get_total_devices(true).first : xrt_core::get_total_devices(false).first;
+  if(devices == 0) 
+    throw std::runtime_error("No devices found");
+  
+  //check if edge, return the first device
+  uint16_t d = 0;
+  auto device = _inUserDomain ? get_userpf_device(d) : get_mgmtpf_device(d);
+  auto bdf = device_query<query::pcie_bdf>(device);
+  if(std::get<0>(bdf) == 0 && std::get<1>(bdf) == 0 && std::get<2>(bdf) == 0) {
+    return 0;
+  }
+
+  // check if we have 2-3 tokens: domain, bus, device.function
+  // domain is optional
+  if(tokens.size() == 2 || tokens.size() == 3) {
```

Comment:
This code it too complicated.  This is because the first element (e.g., "domain") is optional.   

If the token vector was reversed (e.g., `std::revers(std::begin(tokens), std::end(tokens));`)  it will become more manageable.  Where:
index 0 is device.function
index 1 is bus
index 2 is domain (if available)

Doing so removes all of this "special" indexing being performed below.

Reply:
Done